### DISCUSSION TITLE
MonitorControl v4.0.0-beta2

### DISCUSSION THREAD

**Discussion by JoniVR** [MonitorControl/MonitorControl#681] [Release Discussion]
### Enhancements

- Added Internet Access Policy
- Don't relinquish control over brightness keys with no external display connected if fine brightness OSD scale is active
- Changed icon order in menu (when icon mode is enabled)
- Gear shape icon is used for preferences + stands a little bit apart to help user focus.
- Changed default to a minimum software dimming of 15% for safety reasons.
- Added advanced option to enable zero brightness with software dimming.
- Added 'Avoid gamma table manipulation' option for coexistence with f.lux.
- Set relevant options to disabled when keyboard control is disabled.
- Added <kbd>Command</kbd> + <kbd>Q</kbd> shortcut in menu when it is in standard text mode (not icon mode).
- Made preferences more spacious + more room for verbose languages

### Updated translations:

- English - base language
- Chinese (Traditional, Taiwan) - thanks to @stiiveo, @dororojames
- Dutch - thanks to @JoniVR
- French - thanks to @the0neyouseek
- German - thanks to @jajoho
- Hungarian - thanks to @waydabber
- Italian - thanks to @picov
- Korean - thanks to @zzulu
- Turkish - thanks to @mennan

### Under the hood changes & fixes

- Register DDC command touched status. When write on startup enabled, apply only touched command values.
- Reorganised PrefKey list to be less confusing.
- Disengage custom shortcut keyboard after 100 key repeat to prevent possibly endless loop if keyUp event never arrives due to any circumstance.
- Fixed text for external display brightness control keyboard shortcut.
- Fix cumulative darkening bug upon toggling 'Disable dimming as fallback'.
- Make sure that key repeat speed for custom shortcuts do not go below a certain threshold.
- Fixes text clipping issues for various languages
- Added beta channel update backend
- Fixed layout issue at brightness custom shortcuts.
- Fixed custom key shortcuts going runaway when menu was opened during a key repeat streak.

### Known issues

- Some preference panes may not fit to low resolution screens (1280x800 or less) in Advanced mode.

### Notes
- This is a beta release. If you encounter any issues, you can try the [last stable release v3.1.1](https://github.com/MonitorControl/MonitorControl/releases/tag/v3.1.1) or the previous beta. 
- If you are new to v4.0.0, please take a look at the [changelog of v4.0.0-beta1](https://github.com/MonitorControl/MonitorControl/releases/tag/v4.0.0-beta1) for the list of major new features!
- If you discover issues, make sure to report them in the linked release discussion.
- If you want to help translating, see #637
- Thanks to everyone who helps make MonitorControl better!

<hr /><em>This discussion was created from the release <a href='https://github.com/MonitorControl/MonitorControl/releases/tag/v4.0.0-beta2'>MonitorControl v4.0.0-beta2</a>.</em>

**Comment by jajoho** [comment]
Great update!
However, two small things still caught my eye:
1. The update description is loaded as a full website and not as an enumeration or similar.
<img width="652" alt="Bildschirmfoto 2021-10-07 um 20 49 26" src="https://user-images.githubusercontent.com/15175599/136445520-30b814ae-e342-4c1b-9048-c5a4de26bd1f.png">

2. With my screen size (MPB 13") the settings window in german is too small and it can not be scrolled
<img width="732" alt="Bildschirmfoto 2021-10-07 um 20 51 43" src="https://user-images.githubusercontent.com/15175599/136445753-c89ba34c-6fda-45be-9513-166778a18895.png">

**Reply by JoniVR** [reply]
I'm working on the changelog in Sparkle issue, had noticed it too, first time doing Sparkle updates so obviously a learning curve 😅

**Reply by JoniVR** [reply]
Update: Should be better now:
![image](https://user-images.githubusercontent.com/7591717/136453957-82f9a14d-3291-4192-8661-927b092d69b7.png)

We will probably still update the style of these a bit, I just put this together real quick as it's a beta anyways, lots of stuff still needs to be improved/automated for Sparkle updates but at least the basics seem to be working.

**Reply by waydabber** [reply]
You are right, even with the English version of the page barely fits on 1280x800 which is the default resolution for a 13" MBP in Advanced mode. I'll figure something out!

**Reply by waydabber** [reply]
I combined the three startup options into a single popup menu to condense the page (and make it more clean). I hope this now fits in German as well.

<img width="842" alt="Screen Shot 2021-10-08 at 21 16 31" src="https://user-images.githubusercontent.com/37590873/136613754-5f4e5046-9c37-4d0f-af7d-4451c2635013.png">
.

**Reply by waydabber** [reply]
(other things could be changed but I do not want to change anything that upsets already translated texts in beta2)

**Comment by jajoho** [comment]
When I search for updates, I would expect that I could also do that in the settings somewhere. Maybe a button could be added under "About"?

**Reply by waydabber** [reply]
The app will check for updates periodically anyway. We could add a 'Check now' button under 'Automatically check for updates', but the General tab is already a bit crowded.

**Reply by jajoho** [reply]
Thats why I thought a button under "About" (where the version number is displayed) could be suitable? 
But this is definitely nothing important. I just thought about this and wanted to write this down before I forget it.

**Comment by Sayvai** [comment]
One question, since i've updated from `v4.0.0-beta1` to `v4.0.0-beta2`, i noticed that whenever i lowered brightness below 50% on my external **LG UltraWide 34UC99-W** monitor, the software keeps bumping up the brightness to 50%, either via Mac Keyboard brightness keys or through menu slider.

Also, even when the brightness jumps to 50%, the level indicators (on-screen display and menu slider brightness position) doesn't reflect the 50% jump.

However, this cannot be replicated on the main MacBook screen, which can still go down and stay at zero brightness.

Is this intentional? Anyway to stop the fixed minimum brightness jump to 50% on external monitors? 🤔

**Reply by waydabber** [reply]
Hey @Sayvai, are you using f.lux by any chance?

**Reply by waydabber** [reply]
Also can you please try two things:

1. If you enable 'Avoid gamma table manipulation' - is the brightness bump up happening?
2. If you move the 'Combined dimming switchover point' slider to let's say about 30%, does the jump to only 30% then?

<img width="842" alt="Screen Shot 2021-10-09 at 10 08 45" src="https://user-images.githubusercontent.com/37590873/136650057-881a7a08-1e00-4622-9697-a8cbf835a4f9.png">

The results would help me understand the nature of the problem. Thank you!

**Reply by Sayvai** [reply]
Hey @waydabber , yes i am indeed using `f.lux` (and for a long while!).

I've temporarily disabled `f.lux` since you've mentioned it, and the brightness, when set at less than 50%, is maintained at the level below 50%.

Is there any way around this to work with `f.lux`?

**Reply by waydabber** [reply]
Yes, if you want to use f.lux, you have to enable `Avoid gamma table manipulation`. An other option is to activate `Disable software dimming as fallback` and also disable `Combine hardware and software dimming` in General (so only hardware DDC dimming works).

**Reply by Sayvai** [reply]
Thanks @waydabber .

The option to `Avoid gamma table manipulation` is disabled however. 🤔 

![Screenshot 2021-10-09 at 09 31 36](https://user-images.githubusercontent.com/7581546/136650909-c42eafcc-f7d2-4134-830a-4d3a4b5e4c54.png)

Given the above, i applied the alternative suggested option; Activate `Disable software dimming as fallback` and disable `Combine hardware and software dimming`, which appears to work with hardware-only DDC dimming, like you've mentioned.

Saying that, the main MacBook screen is still able to go down to zero brightness (black screen, no visible output). While the external monitor still maintains low level visibility when both monitor brightness settings are set to zero. Is that expected? As I would have thought the main MacBook screen would leave minimal / small amount of brightness at level zero, to be consistent with the external monitor.

**Reply by waydabber** [reply]
For the built-in displays the preferred control method is Apple's own protocol so gamma table manipulation is not needed. In a future version maybe a fuller range of custom settings will be offered though.

If combined brightness is disabled, then the external display brightness cannot go lower than what the lowest display backlight setting possible (changed via DDC) which is usually sill much brighter (due to lower quality lighting solutions) than what Apple or built-in displays allow.

If you want the external display to go down to full blank, you have to enable `Combine hardware and software dimming` and also enable `Allow zero brightness...` (in v4.0.0-beta1 allowing this was the default but since this might be confusing to some, we do not allow full black by default since beta2):

<img width="842" alt="Screen Shot 2021-10-09 at 11 41 17" src="https://user-images.githubusercontent.com/37590873/136653091-50c616de-60da-4b08-b2b2-391a5c762cb3.png">

**Reply by Sayvai** [reply]
Fair enough, and that makes sense.

Given i am also using `f.lux`, I don't really have the luxury to activate `Allow zero brightness via software or combined dimming`.

It's fine for now, as the brightness adjustments experience (based on current adapted configurations to work as best as possible with `f.lux`) is the same as it was in version `3`.

Thank you for your assistance in the meantime.

**Reply by waydabber** [reply]
If you activate `Avoid gamma table manipulation` for the external display, then you can turn on combined dimming. The app starts to use an alternative method to software dim the display (but this method has some drawbacks like mouse cursor will not be affected).

**Reply by Sayvai** [reply]
@waydabber - Not sure if you caught my earlier message thoroughly, but that option; `Avoid gamma table manipulation` is not selectable for me for some reason. I assume it's not supported for my device (MacBook, or External Monitor).

![Screenshot 2021-10-09 at 09 31 36](https://user-images.githubusercontent.com/7581546/136650909-c42eafcc-f7d2-4134-830a-4d3a4b5e4c54.png)

----
OS: **Big Sur 11.6** (20G165)
Model: **MacBook Air** (Retina, 13-inch, 2020)
Processor: **1.2 GHz Quad-Core Intel Core i7**
External monitor: **LG UltraWide 34UC99-W**

**Reply by waydabber** [reply]
It is not selectable for built in displays (which is shown on your screnshot), but for the external display it should be selectable. If you have two displays then you should scroll to the external display in this window. :)

**Reply by waydabber** [reply]
As the window content is scrollable (maybe it is not obvious enough from the UI, which is  a problem) if there are more than one displays:

<img width="842" alt="Screen Shot 2021-10-09 at 15 08 28" src="https://user-images.githubusercontent.com/37590873/136659030-9230c9cd-dcb5-4ee7-9db5-f49062d06209.png">

**Reply by Sayvai** [reply]
Aha, yes, i did not realise that window was at all scrollable until now! 

Maybe the UI could be improved to indicate more options, as the scrollbar is only visible when I actually scroll within the window. 

Perhaps, maybe consider a statically-fixed-positioned set of dynamically generated tabs for each connected monitor at the top, underneath the big main icons (i.e. General, App menu, Keyboard, etc), where the tabs could be labelled with the various display names derived from the available connected displays metadata. And when clicked on, would only load the configurations for the selected display. Basically, something like what is suggested below with my mockup image, if possible? 

And also to enable the scrollbar display to be permanent, if it can be calculated for when there is overflowed content exceeding current window size? 🤔

![Screenshot 2021-10-09 at 14 15 58](https://user-images.githubusercontent.com/7581546/136659586-f5e6e049-5462-4d38-a6a9-bc8d64f78834.png)

Other than that, i've successfully activated `Avoid gamma table manipulation` for the external monitor, and am now able to experience a very-close-to-full-blackout on the external monitor too, just like my in-built display, no minimum brightness threshold jumps this time, and with `f.lux` running at the same time! 

Many thanks @waydabber 💯

**Reply by waydabber** [reply]
Yes, the extra content should be more obvious. I'll simply add a permanent scroll bar to indicate that there is more.

<img width="842" alt="Screen Shot 2021-10-09 at 16 01 40" src="https://user-images.githubusercontent.com/37590873/136661023-204e0791-0a6c-4f1e-b08a-e4094dfdf2ba.png">
